File: app/src/main/java/piuk/blockchain/android/ui/account/AccountActivity.java
package piuk.blockchain.android.ui.account;

import com.google.zxing.BarcodeFormat;

import android.Manifest;
import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.AppCompatEditText;
import android.support.v7.widget.LinearLayoutManager;
import android.text.InputFilter;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.CheckBox;

import info.blockchain.wallet.payload.data.Account;
import info.blockchain.wallet.payload.data.LegacyAddress;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

import javax.inject.Inject;

import piuk.blockchain.android.R;
import piuk.blockchain.android.data.connectivity.ConnectivityStatus;
import piuk.blockchain.android.databinding.ActivityAccountsBinding;
import piuk.blockchain.android.injection.Injector;
import piuk.blockchain.android.ui.account.adapter.AccountAdapter;
import piuk.blockchain.android.ui.account.adapter.AccountHeadersListener;
import piuk.blockchain.android.ui.backup.transfer.ConfirmFundsTransferDialogFragment;
import piuk.blockchain.android.ui.balance.BalanceFragment;
import piuk.blockchain.android.ui.base.BaseMvpActivity;
import piuk.blockchain.android.ui.customviews.MaterialProgressDialog;
import piuk.blockchain.android.ui.customviews.ToastCustom;
import piuk.blockchain.android.ui.zxing.CaptureActivity;
import piuk.blockchain.android.ui.zxing.Intents;
import piuk.blockchain.android.util.MonetaryUtil;
import piuk.blockchain.android.util.PermissionUtil;
import piuk.blockchain.android.util.PrefsUtil;
import piuk.blockchain.android.util.ViewUtils;
import piuk.blockchain.android.util.annotations.Thunk;

import static piuk.blockchain.android.ui.account.AccountPresenter.KEY_WARN_TRANSFER_ALL;

public class AccountActivity extends BaseMvpActivity<AccountView, AccountPresenter>
        implements AccountView {

    private static final int IMPORT_PRIVATE_REQUEST_CODE = 2006;
    private static final int EDIT_ACTIVITY_REQUEST_CODE = 2007;
    private static final int ADDRESS_LABEL_MAX_LENGTH = 17;

    @Inject AccountPresenter accountPresenter;

    private BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, final Intent intent) {
            if (BalanceFragment.ACTION_INTENT.equals(intent.getAction())) {
                onUpdateAccountsList();
                // Check if we need to hide/show the transfer funds icon in the Toolbar
                getPresenter().checkTransferableLegacyFunds(false, false);
            }
        }
    };
    private ArrayList<AccountItem> accountsAndImportedList;
    private AccountAdapter accountsAdapter;
    private List<LegacyAddress> legacy;
    private MenuItem transferFundsMenuItem;
    private PrefsUtil prefsUtil;
    private MonetaryUtil monetaryUtil;
    @Thunk MaterialProgressDialog progress;
    @Thunk ActivityAccountsBinding binding;

    {
        Injector.getInstance().getPresenterComponent().inject(this);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        prefsUtil = new PrefsUtil(this);
        monetaryUtil = new MonetaryUtil(prefsUtil.getValue(PrefsUtil.KEY_BTC_UNITS, MonetaryUtil.UNIT_BTC));

        binding = DataBindingUtil.setContentView(this, R.layout.activity_accounts);

        setupToolbar(binding.toolbarContainer.toolbarGeneral, R.string.addresses);

        binding.accountsList.setLayoutManager(new AccountLayoutManager());
        binding.accountsList.setHasFixedSize(true);
        accountsAndImportedList = new ArrayList<>();

        onUpdateAccountsList();
    }

    @Thunk
    void onRowClick(int position) {
        Intent intent = new Intent(this, AccountEditActivity.class);
        if (position >= getPresenter().getAccounts().size()) {
            intent.putExtra("address_index", position - getPresenter().getAccounts().size());
        } else {
            intent.putExtra("account_index", position);
        }
        startActivityForResult(intent, EDIT_ACTIVITY_REQUEST_CODE);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_account, menu);
        transferFundsMenuItem = menu.findItem(R.id.action_transfer_funds);
        getPresenter().checkTransferableLegacyFunds(true, true);//Auto popup

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            case R.id.action_transfer_funds:
                showProgressDialog(R.string.please_wait);
                getPresenter().checkTransferableLegacyFunds(false, true);//Not auto popup
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void startScanForResult() {
        Intent intent = new Intent(AccountActivity.this, CaptureActivity.class);
        intent.putExtra(Intents.Scan.FORMATS, EnumSet.allOf(BarcodeFormat.class));
        intent.putExtra(Intents.Scan.MODE, Intents.Scan.QR_CODE_MODE);
        startActivityForResult(intent, IMPORT_PRIVATE_REQUEST_CODE);
    }

    @Thunk
    void importAddress() {
        if (ContextCompat.checkSelfPermission(AccountActivity.this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            PermissionUtil.requestCameraPermissionFromActivity(binding.mainLayout, AccountActivity.this);
        } else {
            onScanButtonClicked();
        }
    }

    private void onScanButtonClicked() {
        new SecondPasswordHandler(this).validate(new SecondPasswordHandler.ResultListener() {
            @Override
            public void onNoSecondPassword() {
                getPresenter().onScanButtonClicked();
            }

            @Override
            public void onSecondPasswordValidated(String validateSecondPassword) {
                getPresenter().setDoubleEncryptionPassword(validateSecondPassword);
                getPresenter().onScanButtonClicked();
            }
        });
    }

    @Thunk
    void createNewAccount() {
        new SecondPasswordHandler(this).validate(new SecondPasswordHandler.ResultListener() {
            @Override
            public void onNoSecondPassword() {
                promptForAccountLabel();
            }

            @Override
            public void onSecondPasswordValidated(String validateSecondPassword) {
                getPresenter().setDoubleEncryptionPassword(validateSecondPassword);
                promptForAccountLabel();
            }
        });
    }

    @Thunk
    void promptForAccountLabel() {
        AppCompatEditText editText = getAddressLabelEditText();

        new AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.label)
                .setMessage(R.string.assign_display_name)
                .setView(ViewUtils.getAlertDialogPaddedView(this, editText))
                .setCancelable(false)
                .setPositiveButton(R.string.save_name, (dialog, whichButton) -> {
                    if (!editText.getText().toString().trim().isEmpty()) {
                        addAccount(editText.getText().toString().trim());
                    } else {
                        ToastCustom.makeText(AccountActivity.this, getResources().getString(R.string.label_cant_be_empty), ToastCustom.LENGTH_SHORT, ToastCustom.TYPE_ERROR);
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    private void addAccount(final String accountLabel) {
        if (!ConnectivityStatus.hasConnectivity(AccountActivity.this)) {
            ToastCustom.makeText(AccountActivity.this, getString(R.string.check_connectivity_exit), ToastCustom.LENGTH_SHORT, ToastCustom.TYPE_ERROR);
        } else {
            getPresenter().createNewAccount(accountLabel);
        }
    }

    @Override
    public void onUpdateAccountsList() {
//        accountsAndImportedList is linked to AccountAdapter - do not reconstruct or loose reference otherwise notifyDataSetChanged won't work
        accountsAndImportedList.clear();
        int correctedPosition = 0;
        List<Account> accounts = getPresenter().getAccounts();
        List<Account> accountClone = new ArrayList<>(accounts.size());
        accountClone.addAll(accounts);

        // Create New Wallet button at top position
        accountsAndImportedList.add(new AccountItem(AccountItem.TYPE_CREATE_NEW_WALLET_BUTTON));

        int defaultIndex = getPresenter().getDefaultAccountIndex();
        Account defaultAccount = getPresenter().getAccounts().get(defaultIndex);

        for (int i = 0; i < accountClone.size(); i++) {
            String label = accountClone.get(i).getLabel();
            String balance = getAccountBalance(i);

            if (label != null && label.length() > ADDRESS_LABEL_MAX_LENGTH) {
                label = label.substring(0, ADDRESS_LABEL_MAX_LENGTH) + "...";
            }
            if (label == null || label.isEmpty()) label = "";

            accountsAndImportedList.add(
                    new AccountItem(
                            correctedPosition,
                            label,
                            null,
                            balance,
                            accountClone.get(i).isArchived(),
                            false,
                            defaultAccount.getXpub().equals(accountClone.get(i).getXpub()),
                            AccountItem.TYPE_ACCOUNT));
            correctedPosition++;
        }

        // Import Address button at first position after wallets
        accountsAndImportedList.add(new AccountItem(AccountItem.TYPE_IMPORT_ADDRESS_BUTTON));

        legacy = getPresenter().getLegacyAddressList();
        for (int j = 0; j < legacy.size(); j++) {

            String label = legacy.get(j).getLabel();
            String address = legacy.get(j).getAddress();
            String balance = getAddressBalance(j);

            if (label != null && label.length() > ADDRESS_LABEL_MAX_LENGTH)
                label = label.substring(0, ADDRESS_LABEL_MAX_LENGTH) + "...";
            if (label == null || label.isEmpty()) label = "";
            if (address == null || address.isEmpty()) address = "";

            accountsAndImportedList.add(
                    new AccountItem(
                            correctedPosition,
                            label,
                            address,
                            balance,
                            legacy.get(j).getTag() == LegacyAddress.ARCHIVED_ADDRESS,
                            legacy.get(j).isWatchOnly(),
                            false,
                            AccountItem.TYPE_ACCOUNT));
            correctedPosition++;
        }

        if (accountsAdapter == null) {
            accountsAdapter = new AccountAdapter(new AccountHeadersListener() {
                @Override
                public void onCreateNewClicked() {
                    createNewAccount();
                }

                @Override
                public void onImportAddressClicked() {
                    importAddress();
                }

                @Override
                public void onAccountClicked(int correctedPosition) {
                    onRowClick(correctedPosition);
                }
            });

            accountsAdapter.setItems(accountsAndImportedList);
            binding.accountsList.setAdapter(accountsAdapter);
        } else {
            // Notify adapter of items changes
            accountsAdapter.notifyDataSetChanged();
        }
    }

    private String getAccountBalance(int index) {
        String address = getPresenter().getXpubFromIndex(index);
        BigInteger addressBalance = getPresenter().getBalanceFromAddress(address);
        // Archived addresses/xPubs aren't parsed, so balance will be null
        Long amount = addressBalance != null ? addressBalance.longValue() : 0L;

        String unit = monetaryUtil.getBtcUnits()[prefsUtil.getValue(PrefsUtil.KEY_BTC_UNITS, MonetaryUtil.UNIT_BTC)];

        return monetaryUtil.getDisplayAmount(amount) + " " + unit;
    }

    private String getAddressBalance(int index) {
        String address = legacy.get(index).getAddress();
        Long amount = getPresenter().getBalanceFromAddress(address).longValue();
        String unit = monetaryUtil.getBtcUnits()[prefsUtil.getValue(PrefsUtil.KEY_BTC_UNITS, MonetaryUtil.UNIT_BTC)];

        return monetaryUtil.getDisplayAmount(amount) + " " + unit;
    }

    @Override
    protected void onResume() {
        super.onResume();
        IntentFilter filter = new IntentFilter(BalanceFragment.ACTION_INTENT);
        LocalBroadcastManager.getInstance(AccountActivity.this).registerReceiver(receiver, filter);
        onUpdateAccountsList();
    }

    @Override
    public void onPause() {
        LocalBroadcastManager.getInstance(AccountActivity.this).unregisterReceiver(receiver);
        super.onPause();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK && requestCode == IMPORT_PRIVATE_REQUEST_CODE
                && data != null && data.getStringExtra(CaptureActivity.SCAN_RESULT) != null) {

            String strResult = data.getStringExtra(CaptureActivity.SCAN_RESULT);
            getPresenter().onAddressScanned(strResult);
            setResult(resultCode);
        } else if (resultCode == Activity.RESULT_OK && requestCode == EDIT_ACTIVITY_REQUEST_CODE) {
            onUpdateAccountsList();
            setResult(resultCode);
        }
    }

    @Override
    public void showBip38PasswordDialog(String data) {
        AppCompatEditText password = new AppCompatEditText(this);
        password.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);

        new AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.app_name)
                .setMessage(R.string.bip38_password_entry)
                .setView(ViewUtils.getAlertDialogPaddedView(this, password))
                .setCancelable(false)
                .setPositiveButton(android.R.string.ok, (dialog, whichButton) ->
                        getPresenter().importBip38Address(data, password.getText().toString()))
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    @Override
    public void showWatchOnlyWarningDialog(String address) {
        new AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.warning)
                .setCancelable(false)
                .setMessage(getString(R.string.watch_only_import_warning))
                .setPositiveButton(R.string.dialog_continue, (dialog, whichButton) -> getPresenter().confirmImportWatchOnly(address))
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    @Override
    public void showRenameImportedAddressDialog(LegacyAddress address) {
        AppCompatEditText editText = getAddressLabelEditText();

        new AlertDialog.Builder(AccountActivity.this, R.style.AlertDialogStyle)
                .setTitle(R.string.app_name)
                .setMessage(R.string.label_address)
                .setView(ViewUtils.getAlertDialogPaddedView(this, editText))
                .setCancelable(false)
                .setPositiveButton(R.string.save_name, (dialog, whichButton) -> {
                    String label = editText.getText().toString();
                    if (!label.trim().isEmpty()) {
                        address.setLabel(label);
                    }

                    remoteSaveNewAddress(address);
                })
                .setNegativeButton(R.string.polite_no, (dialog, whichButton) -> {
                    remoteSaveNewAddress(address);
                }).show();
    }

    @Override
    public void showToast(@StringRes int message, @ToastCustom.ToastType String toastType) {
        ToastCustom.makeText(this, getString(message), ToastCustom.LENGTH_SHORT, toastType);
    }

    @Override
    public void broadcastIntent(Intent intent) {
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
    }

    private void remoteSaveNewAddress(final LegacyAddress legacy) {
        if (!ConnectivityStatus.hasConnectivity(AccountActivity.this)) {
            ToastCustom.makeText(AccountActivity.this, getString(R.string.check_connectivity_exit), ToastCustom.LENGTH_SHORT, ToastCustom.TYPE_ERROR);
        } else {
            getPresenter().updateLegacyAddress(legacy);
        }
    }

    @Override
    public void onShowTransferableLegacyFundsWarning(boolean isAutoPopup) {
        CheckBox checkBox = new CheckBox(this);
        checkBox.setChecked(false);
        checkBox.setText(R.string.dont_ask_again);

        AlertDialog.Builder builder = new AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.transfer_funds)
                .setMessage(getString(R.string.transfer_recommend) + "\n")
                .setPositiveButton(R.string.transfer, (dialog, which) -> {
                    transferSpendableFunds();
                    if (checkBox.isChecked()) {
                        prefsUtil.setValue(KEY_WARN_TRANSFER_ALL, false);
                    }
                })
                .setNegativeButton(R.string.not_now, (dialog, which) -> {
                    if (checkBox.isChecked()) {
                        prefsUtil.setValue(KEY_WARN_TRANSFER_ALL, false);
                    }
                });

        if (isAutoPopup) {
            builder.setView(ViewUtils.getAlertDialogPaddedView(this, checkBox));
        }

        AlertDialog alertDialog = builder.create();
        if (!isFinishing()) {
            alertDialog.show();
        }

        Button negative = alertDialog.getButton(DialogInterface.BUTTON_NEGATIVE);
        negative.setTextColor(ContextCompat.getColor(this, R.color.primary_gray_dark));
    }

    private void transferSpendableFunds() {
        ConfirmFundsTransferDialogFragment fragment = ConfirmFundsTransferDialogFragment.newInstance();
        fragment.show(getSupportFragmentManager(), ConfirmFundsTransferDialogFragment.TAG);
    }

    @Override
    public void onSetTransferLegacyFundsMenuItemVisible(boolean visible) {
        transferFundsMenuItem.setVisible(visible);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == PermissionUtil.PERMISSION_REQUEST_CAMERA) {
            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                onScanButtonClicked();
            } else {
                // Permission request was denied.
            }
        } else {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    @NonNull
    private AppCompatEditText getAddressLabelEditText() {
        AppCompatEditText editText = new AppCompatEditText(this);
        editText.setInputType(InputType.TYPE_TEXT_FLAG_CAP_SENTENCES);
        editText.setFilters(new InputFilter[]{new InputFilter.LengthFilter(ADDRESS_LABEL_MAX_LENGTH)});
        editText.setHint(R.string.name);
        return editText;
    }

    @Override
    public void showProgressDialog(@StringRes int message) {
        dismissProgressDialog();
        if (!isFinishing()) {
            progress = new MaterialProgressDialog(this);
            progress.setMessage(message);
            progress.setCancelable(false);
            progress.show();
        }
    }

    @Override
    public void dismissProgressDialog() {
        if (progress != null && progress.isShowing()) {
            progress.dismiss();
            progress = null;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        dismissProgressDialog();
    }

    @Override
    protected AccountPresenter createPresenter() {
        return accountPresenter;
    }

    @Override
    protected AccountView getView() {
        return this;
    }

    private class AccountLayoutManager extends LinearLayoutManager {

        AccountLayoutManager() {
            super(AccountActivity.this);
        }

        @Override
        public boolean supportsPredictiveItemAnimations() {
            return false;
        }
    }

}

File: app/src/main/java/piuk/blockchain/android/ui/account/AccountActivity.kt
package piuk.blockchain.android.ui.account

import android.Manifest
import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.graphics.Rect
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.v4.content.ContextCompat
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AlertDialog
import android.support.v7.widget.AppCompatEditText
import android.support.v7.widget.LinearLayoutManager
import android.text.InputFilter
import android.text.InputType
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import android.widget.CheckBox
import com.google.zxing.BarcodeFormat
import info.blockchain.wallet.payload.data.LegacyAddress
import kotlinx.android.synthetic.main.activity_accounts.*
import kotlinx.android.synthetic.main.toolbar_general.*
import piuk.blockchain.android.R
import piuk.blockchain.android.data.currency.CryptoCurrencies
import piuk.blockchain.android.injection.Injector
import piuk.blockchain.android.ui.account.AccountPresenter.Companion.ADDRESS_LABEL_MAX_LENGTH
import piuk.blockchain.android.ui.account.AccountPresenter.Companion.KEY_WARN_TRANSFER_ALL
import piuk.blockchain.android.ui.account.adapter.AccountAdapter
import piuk.blockchain.android.ui.account.adapter.AccountHeadersListener
import piuk.blockchain.android.ui.backup.transfer.ConfirmFundsTransferDialogFragment
import piuk.blockchain.android.ui.balance.BalanceFragment
import piuk.blockchain.android.ui.base.BaseMvpActivity
import piuk.blockchain.android.ui.customviews.MaterialProgressDialog
import piuk.blockchain.android.ui.customviews.ToastCustom
import piuk.blockchain.android.ui.zxing.CaptureActivity
import piuk.blockchain.android.ui.zxing.Intents
import piuk.blockchain.android.util.PermissionUtil
import piuk.blockchain.android.util.PrefsUtil
import piuk.blockchain.android.util.ViewUtils
import piuk.blockchain.android.util.extensions.getTextString
import piuk.blockchain.android.util.extensions.toast
import piuk.blockchain.android.util.helperfunctions.consume
import piuk.blockchain.android.util.helperfunctions.unsafeLazy
import java.util.*
import javax.inject.Inject

class AccountActivity : BaseMvpActivity<AccountView, AccountPresenter>(), AccountView,
    AccountHeadersListener {

    @Suppress("MemberVisibilityCanBePrivate")
    @Inject lateinit var accountPresenter: AccountPresenter

    override val locale: Locale = Locale.getDefault()

    private val prefsUtil: PrefsUtil by unsafeLazy { PrefsUtil(this) }
    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (BalanceFragment.ACTION_INTENT == intent.action) {
                onViewReady()
                // Check if we need to hide/show the transfer funds icon in the Toolbar
                presenter.checkTransferableLegacyFunds(false, false)
            }
        }
    }

    private lateinit var transferFundsMenuItem: MenuItem
    private val accountsAdapter: AccountAdapter by unsafeLazy { AccountAdapter(this) }
    private var progress: MaterialProgressDialog? = null

    init {
        Injector.getInstance().presenterComponent.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_accounts)
        setupToolbar(toolbar_general, R.string.drawer_addresses)

        with(currency_header) {
            hideEthereum()
            setCurrentlySelectedCurrency(presenter.cryptoCurrency)
            setSelectionListener { presenter.cryptoCurrency = it }
        }

        with(recyclerview_accounts) {
            layoutManager = LinearLayoutManager(this@AccountActivity)
            itemAnimator = null
            setHasFixedSize(true)
            adapter = accountsAdapter
        }

        onViewReady()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_account, menu)
        transferFundsMenuItem = menu.findItem(R.id.action_transfer_funds)
        // Auto popup
        presenter.checkTransferableLegacyFunds(true, true)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> consume { onBackPressed() }
        R.id.action_transfer_funds -> consume {
            showProgressDialog(R.string.please_wait)
            // Not auto popup
            presenter.checkTransferableLegacyFunds(false, true)
        }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (currency_header.isOpen()) {
            currency_header.close()
        } else {
            super.onBackPressed()
        }
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        if (ev.action == MotionEvent.ACTION_DOWN) {
            // Notify touchOutsideViewListeners if user tapped outside a given view
            if (currency_header != null) {
                val viewRect = Rect()
                currency_header.getGlobalVisibleRect(viewRect)
                if (!viewRect.contains(ev.rawX.toInt(), ev.rawY.toInt())) {
                    if (currency_header.isOpen()) {
                        currency_header.close()
                        return false
                    }
                }
            }
        }
        return super.dispatchTouchEvent(ev)
    }

    override fun startScanForResult() {
        Intent(this, CaptureActivity::class.java).apply {
            putExtra(Intents.Scan.FORMATS, EnumSet.allOf(BarcodeFormat::class.java))
            putExtra(Intents.Scan.MODE, Intents.Scan.QR_CODE_MODE)
        }.run { startActivityForResult(this, IMPORT_PRIVATE_REQUEST_CODE) }
    }

    override fun onCreateNewClicked() {
        createNewAccount()
    }

    override fun onImportAddressClicked() {
        importAddress()
    }

    override fun onAccountClicked(cryptoCurrency: CryptoCurrencies, correctedPosition: Int) {
        onRowClick(cryptoCurrency, correctedPosition)
    }

    private fun importAddress() {
        if (ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.CAMERA
            ) != PackageManager.PERMISSION_GRANTED) {
            PermissionUtil.requestCameraPermissionFromActivity(linear_layout_root, this)
        } else {
            onScanButtonClicked()
        }
    }

    private fun onRowClick(cryptoCurrency: CryptoCurrencies, position: Int) {
        AccountEditActivity.startForResult(
                this,
                getAccountPosition(cryptoCurrency, position),
                if (position >= presenter.accountSize) position - presenter.accountSize else -1,
                cryptoCurrency,
                EDIT_ACTIVITY_REQUEST_CODE
        )
    }

    private fun getAccountPosition(cryptoCurrency: CryptoCurrencies, position: Int): Int {
        return if (cryptoCurrency == CryptoCurrencies.BTC) {
            if (position < presenter.accountSize) position else -1
        } else {
            return position
        }
    }

    private fun onScanButtonClicked() {
        SecondPasswordHandler(this).validate(object : SecondPasswordHandler.ResultListener {
            override fun onNoSecondPassword() {
                presenter.onScanButtonClicked()
            }

            override fun onSecondPasswordValidated(validateSecondPassword: String) {
                presenter.doubleEncryptionPassword = validateSecondPassword
                presenter.onScanButtonClicked()
            }
        })
    }

    private fun createNewAccount() {
        SecondPasswordHandler(this).validate(object : SecondPasswordHandler.ResultListener {
            override fun onNoSecondPassword() {
                promptForAccountLabel()
            }

            override fun onSecondPasswordValidated(validateSecondPassword: String) {
                presenter.doubleEncryptionPassword = validateSecondPassword
                promptForAccountLabel()
            }
        })
    }

    private fun promptForAccountLabel() {
        val editText = AppCompatEditText(this).apply {
            inputType = InputType.TYPE_TEXT_FLAG_CAP_SENTENCES
            filters = arrayOf<InputFilter>(InputFilter.LengthFilter(ADDRESS_LABEL_MAX_LENGTH))
            setHint(R.string.name)
        }

        AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.label)
                .setMessage(R.string.assign_display_name)
                .setView(ViewUtils.getAlertDialogPaddedView(this, editText))
                .setCancelable(false)
                .setPositiveButton(R.string.save_name) { _, _ ->
                    if (!editText.getTextString().trim { it <= ' ' }.isEmpty()) {
                        addAccount(editText.getTextString().trim { it <= ' ' })
                    } else {
                        toast(R.string.label_cant_be_empty, ToastCustom.TYPE_ERROR)
                    }
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
    }

    private fun addAccount(accountLabel: String) {
        presenter.createNewAccount(accountLabel)
    }

    override fun updateAccountList(displayAccounts: List<AccountItem>) {
        accountsAdapter.items = displayAccounts
    }

    override fun onResume() {
        super.onResume()
        val filter = IntentFilter(BalanceFragment.ACTION_INTENT)
        LocalBroadcastManager.getInstance(this).registerReceiver(receiver, filter)
        onViewReady()
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(receiver)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK
            && requestCode == IMPORT_PRIVATE_REQUEST_CODE
            && data != null && data.getStringExtra(CaptureActivity.SCAN_RESULT) != null) {

            val strResult = data.getStringExtra(CaptureActivity.SCAN_RESULT)
            presenter.onAddressScanned(strResult)
            setResult(resultCode)
        } else if (resultCode == Activity.RESULT_OK && requestCode == EDIT_ACTIVITY_REQUEST_CODE) {
            onViewReady()
            setResult(resultCode)
        }
    }

    override fun showBip38PasswordDialog(data: String) {
        val password = AppCompatEditText(this).apply {
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
        }

        AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.app_name)
                .setMessage(R.string.bip38_password_entry)
                .setView(ViewUtils.getAlertDialogPaddedView(this, password))
                .setCancelable(false)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    presenter.importBip38Address(data, password.getTextString())
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
    }

    override fun showWatchOnlyWarningDialog(address: String) {
        AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.warning)
                .setCancelable(false)
                .setMessage(getString(R.string.watch_only_import_warning))
                .setPositiveButton(R.string.dialog_continue) { _, _ ->
                    presenter.confirmImportWatchOnly(address)
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
    }

    override fun showRenameImportedAddressDialog(address: LegacyAddress) {
        val editText = AppCompatEditText(this).apply {
            inputType = InputType.TYPE_TEXT_FLAG_CAP_SENTENCES
            filters = arrayOf<InputFilter>(InputFilter.LengthFilter(ADDRESS_LABEL_MAX_LENGTH))
            setHint(R.string.name)
        }

        AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.app_name)
                .setMessage(R.string.label_address)
                .setView(ViewUtils.getAlertDialogPaddedView(this, editText))
                .setCancelable(false)
                .setPositiveButton(R.string.save_name) { _, _ ->
                    val label = editText.getTextString()
                    if (!label.trim { it <= ' ' }.isEmpty()) {
                        address.label = label
                    }

                    remoteSaveNewAddress(address)
                }
                .setNegativeButton(R.string.polite_no) { _, _ -> remoteSaveNewAddress(address) }
                .show()
    }

    override fun showToast(@StringRes message: Int, @ToastCustom.ToastType toastType: String) {
        toast(message, toastType)
    }

    override fun broadcastIntent(intent: Intent) {
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent)
    }

    private fun remoteSaveNewAddress(legacy: LegacyAddress) {
        presenter.updateLegacyAddress(legacy)
    }

    override fun onShowTransferableLegacyFundsWarning(isAutoPopup: Boolean) {
        val checkBox = CheckBox(this)
        checkBox.isChecked = false
        checkBox.setText(R.string.dont_ask_again)

        val builder = AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.transfer_funds)
                .setMessage(getString(R.string.transfer_recommend) + "\n")
                .setPositiveButton(R.string.transfer) { _, _ ->
                    transferSpendableFunds()
                    if (checkBox.isChecked) {
                        prefsUtil.setValue(KEY_WARN_TRANSFER_ALL, false)
                    }
                }
                .setNegativeButton(R.string.not_now) { _, _ ->
                    if (checkBox.isChecked) {
                        prefsUtil.setValue(KEY_WARN_TRANSFER_ALL, false)
                    }
                }

        if (isAutoPopup) {
            builder.setView(ViewUtils.getAlertDialogPaddedView(this, checkBox))
        }

        val alertDialog = builder.create()
        if (!isFinishing) {
            alertDialog.show()
        }

        alertDialog.getButton(DialogInterface.BUTTON_NEGATIVE).apply {
            setTextColor(ContextCompat.getColor(this@AccountActivity, R.color.primary_gray_dark))
        }
    }

    private fun transferSpendableFunds() {
        ConfirmFundsTransferDialogFragment.newInstance()
                .show(supportFragmentManager, ConfirmFundsTransferDialogFragment.TAG)
    }

    override fun onSetTransferLegacyFundsMenuItemVisible(visible: Boolean) {
        transferFundsMenuItem.isVisible = visible
    }

    override fun onRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<String>,
            grantResults: IntArray
    ) {
        if (requestCode == PermissionUtil.PERMISSION_REQUEST_CAMERA) {
            if (grantResults.size == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                onScanButtonClicked()
            } else {
                // Permission request was denied.
            }
        } else {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    override fun showProgressDialog(@StringRes message: Int) {
        dismissProgressDialog()
        if (!isFinishing) {
            progress = MaterialProgressDialog(this).apply {
                setMessage(message)
                setCancelable(false)
                show()
            }
        }
    }

    override fun dismissProgressDialog() {
        if (progress?.isShowing == true) {
            progress!!.dismiss()
            progress = null
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        dismissProgressDialog()
    }

    override fun createPresenter() = accountPresenter

    override fun getView() = this

    companion object {

        private const val IMPORT_PRIVATE_REQUEST_CODE = 2006
        private const val EDIT_ACTIVITY_REQUEST_CODE = 2007

    }

}
